overlord/snapstate/snapstate.go

```diff
@@ -2301,15 +2355,57 @@ func Update(st *state.State, name string, opts *RevisionOptions, userID int, fla
 	return UpdateWithDeviceContext(st, name, opts, userID, flags, nil, "")
 }
 
+type snapInfoForUpdate func(dc DeviceContext, ro *RevisionOptions, fl Flags, snapst *SnapState) ([]minimalInstallInfo, error)
+
 // UpdateWithDeviceContext initiates a change updating a snap.
-// It will query for the snap with the given deviceCtx.
+// It will query the store for the snap with the given deviceCtx.
 // Note that the state must be locked by the caller.
 //
 // The returned TaskSet can contain a LastBeforeLocalModificationsEdge
 // identifying the last task before the first task that introduces system
 // modifications. If no such edge is set, then none of the tasks introduce
 // system modifications.
 func UpdateWithDeviceContext(st *state.State, name string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext, fromChange string) (*state.TaskSet, error) {
+	snapUpdateInfo := func(dc DeviceContext, ro *RevisionOptions, fl Flags, snapst *SnapState) ([]minimalInstallInfo, error) {
+		toUpdate := []minimalInstallInfo{}
+		info, infoErr := infoForUpdate(st, snapst, name, ro, userID, fl, dc)
+		switch infoErr {
+		case nil:
+			toUpdate = append(toUpdate, installSnapInfo{info})
+		case store.ErrNoUpdateAvailable:
+			// there may be some new auto-aliases
+		default:
+			return nil, infoErr
+		}
+		return toUpdate, infoErr
+	}
+
+	return updateWithDeviceContext(st, name, opts, userID, flags, deviceCtx, fromChange, snapUpdateInfo)
+}
+
+// UpdatePathWithDeviceContext initiates a change updating a snap from a local file.
+// Note that the state must be locked by the caller.
+//
+// The returned TaskSet can contain a LastBeforeLocalModificationsEdge
+// identifying the last task before the first task that introduces system
+// modifications. If no such edge is set, then none of the tasks introduce
+// system modifications.
+func UpdatePathWithDeviceContext(st *state.State, name string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext, fromChange string, si *snap.SideInfo, path string) (*state.TaskSet, error) {
```

Comment:
this does not seem tested also do we need a way to trigger the code paths corresponding to store.ErrNoUpdateAvailable as it can be returned in the store case, but here we do not have a store? those allow channel switching etc

Reply:
Changed